File: 3-APIs/kaffebar-java-solution/src/main/java/no/soprasteria/kaffebar/events/OrderEventPublisher.java
package no.soprasteria.kaffebar.events;

// Spring Boot 4 leverer Jackson 3 (tools.jackson) som JSON-mapper. Den genererte
// koden bruker fortsatt Jackson 2-annotasjonene, som Jackson 3 leser som før.
import tools.jackson.databind.ObjectMapper;
import no.soprasteria.kaffebar.model.Health;
import no.soprasteria.kaffebar.model.Order;
import no.soprasteria.kaffebar.model.OrderStatus;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.amqp.core.Message;
import org.springframework.amqp.core.MessageDeliveryMode;
import org.springframework.amqp.core.MessageProperties;
import org.springframework.amqp.rabbit.connection.Connection;
import org.springframework.amqp.rabbit.connection.ConnectionFactory;
import org.springframework.amqp.rabbit.core.RabbitTemplate;
import org.springframework.beans.factory.ObjectProvider;
import org.springframework.boot.context.event.ApplicationReadyEvent;
import org.springframework.context.event.EventListener;
import org.springframework.stereotype.Component;

import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.time.temporal.ChronoUnit;

/**
 * VALGFRITT SISTE STEG — ikke en del av oppgavesettet i samling 3.
 *
 * Oppgavene sier ingenting om hendelser. Men peker en deltaker frontenden i samling 4
 * på sitt eget API fra mai, får hen bare steg 1 (data over HTTP), ikke steg 2 (live
 * oppdatering i to vinduer samtidig). Derfor ligger publiseringen her — bak
 * {@code kaffebar.events.enabled}, som er {@code false} som standard.
 *
 * Er flagget av, gjør denne klassen ingenting i det hele tatt. Ingen tilkobling,
 * ingen logglinjer, ingen endret oppførsel på HTTP-siden.
 *
 * Samme exchange (`kaffebar`, topic, durable), samme routing keys
 * (`order.created`, `order.status-changed`) og samme meldingsformat som
 * 4-Frontend/kaffebar-api.
 */
@Component
public class OrderEventPublisher {

    private static final Logger log = LoggerFactory.getLogger(OrderEventPublisher.class);

    private final KaffebarProperties properties;
    private final ObjectProvider<RabbitTemplate> rabbitTemplate;
    private final ObjectProvider<ConnectionFactory> connectionFactory;
    private final ObjectMapper objectMapper;

    private volatile boolean connected = false;

    public OrderEventPublisher(KaffebarProperties properties,
                               ObjectProvider<RabbitTemplate> rabbitTemplate,
                               ObjectProvider<ConnectionFactory> connectionFactory,
                               ObjectMapper objectMapper) {
        this.properties = properties;
        this.rabbitTemplate = rabbitTemplate;
        this.connectionFactory = connectionFactory;
        this.objectMapper = objectMapper;
    }

    /**
     * Kobler til ved oppstart, slik at {@code GET /health} kan si sannheten med én gang
     * og exchangen blir deklarert av RabbitAdmin. Kaster aldri: API-et skal starte og
     * fungere selv om broker er nede, akkurat som referanse-implementasjonen.
     */
    @EventListener(ApplicationReadyEvent.class)
    void connect() {
        if (!properties.getEvents().isEnabled()) {
            return;
        }
        try (Connection connection = connectionFactory.getObject().createConnection()) {
            connected = connection.isOpen();
            log.info("Koblet til RabbitMQ. Exchange \"{}\" (topic, durable).",
                    properties.getEvents().getExchange());
        } catch (Exception exception) {
            connected = false;
            log.warn("Får ikke kontakt med RabbitMQ ({}). API-et kjører videre uten hendelser.",
                    exception.getMessage());
        }
    }

    public void orderCreated(Order order) {
        publish("order.created", new KaffebarEvent("order.created", timestamp(), order, null));
    }

    public void orderStatusChanged(Order order, OrderStatus previousStatus) {
        publish("order.status-changed",
                new KaffebarEvent("order.status-changed", timestamp(), order, previousStatus));
    }

    /** Hva {@code GET /health} skal rapportere. */
    public Health.RabbitmqEnum connectionState() {
        if (!properties.getEvents().isEnabled()) {
            return Health.RabbitmqEnum.DISABLED;
        }
        return connected ? Health.RabbitmqEnum.CONNECTED : Health.RabbitmqEnum.DISCONNECTED;
    }

    private void publish(String routingKey, KaffebarEvent event) {
        if (!properties.getEvents().isEnabled()) {
            return;
        }
        try {
            MessageProperties messageProperties = new MessageProperties();
            messageProperties.setContentType(MessageProperties.CONTENT_TYPE_JSON);
            messageProperties.setContentEncoding("UTF-8");
            messageProperties.setDeliveryMode(MessageDeliveryMode.PERSISTENT);

            Message message = new Message(objectMapper.writeValueAsBytes(event), messageProperties);
            rabbitTemplate.getObject().send(properties.getEvents().getExchange(), routingKey, message);
            connected = true;
        } catch (Exception exception) {
            // Hendelser er en bonus. En broker som er nede skal aldri gjøre et
            // HTTP-kall som ellers gikk bra om til en 500.
            connected = false;
            log.warn("Klarte ikke å publisere {}: {}", routingKey, exception.getMessage());
        }
    }

    private static String timestamp() {
        return DateTimeFormatter.ISO_OFFSET_DATE_TIME.format(
                OffsetDateTime.now(ZoneOffset.UTC).truncatedTo(ChronoUnit.MILLIS));
    }
}

File: 3-APIs/kaffebar-dotnet-solution/Kaffebar/Events/OrderEventPublisher.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using Kaffebar.Models;
using RabbitMQ.Client;

namespace Kaffebar.Events;

/// <summary>Meldingsformatet på exchangen.</summary>
/// <remarks>
/// Bit for bit likt det 4-Frontend/kaffebar-api publiserer, slik at
/// <c>/api/events</c>-route handleren i kaffebar-web ikke merker forskjell på hvilken
/// implementasjon som står bak.
/// </remarks>
public record KaffebarEvent(
    string Type,
    string Timestamp,
    Order Data,
    [property: JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    OrderStatus? PreviousStatus);

/// <summary>
/// VALGFRITT SISTE STEG — ikke en del av oppgavesettet i samling 3.
/// </summary>
/// <remarks>
/// Oppgavene sier ingenting om hendelser. Men peker en deltaker frontenden i samling 4
/// på sitt eget API fra mai, får hen bare steg 1 (data over HTTP), ikke steg 2 (to
/// vinduer som oppdaterer seg selv). Uten hendelser mangler halve poenget med
/// samlingen.
///
/// Er <c>Kaffebar:Events:Enabled</c> false — som er standard — gjør denne klassen
/// bokstavelig talt ingenting. Ingen tilkobling, ingen logglinjer, ingen endret
/// HTTP-oppførsel.
///
/// Samme exchange (<c>kaffebar</c>, topic, durable), samme routing keys
/// (<c>order.created</c>, <c>order.status-changed</c>) og samme meldingsformat som
/// 4-Frontend/kaffebar-api.
/// </remarks>
public sealed class OrderEventPublisher(
    EventOptions options,
    IHostApplicationLifetime lifetime,
    ILogger<OrderEventPublisher> logger) : IAsyncDisposable
{
    private static readonly JsonSerializerOptions JsonOptions = new(JsonSerializerDefaults.Web)
    {
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
        // Samme konvertere som HTTP-svarene. Hendelsen skal inneholde nøyaktig det
        // samme Order-objektet som GET /orders/{id} ville gitt — ellers får en
        // konsument to litt ulike sannheter om samme ordre.
        Converters = { new JsonStringEnumConverter(), new UtcDateTimeOffsetConverter() }
    };

    private readonly SemaphoreSlim _gate = new(1, 1);
    private IConnection? _connection;
    private IChannel? _channel;

    public RabbitmqState State =>
        !options.Enabled ? RabbitmqState.Disabled
        : _channel is { IsOpen: true } ? RabbitmqState.Connected
        : RabbitmqState.Disconnected;

    public void OrderCreated(Order order) =>
        Publish("order.created", new KaffebarEvent("order.created", Timestamp(), order, null));

    public void OrderStatusChanged(Order order, OrderStatus previousStatus) =>
        Publish("order.status-changed",
            new KaffebarEvent("order.status-changed", Timestamp(), order, previousStatus));

    /// <summary>
    /// Publiserer i bakgrunnen. HTTP-svaret skal aldri vente på broker, og en broker som
    /// er nede skal aldri gjøre et kall som ellers gikk bra om til en 500. Hendelser er
    /// en bonus; bestillingen er jobben.
    /// </summary>
    private void Publish(string routingKey, KaffebarEvent message)
    {
        if (!options.Enabled)
        {
            return;
        }

        _ = Task.Run(async () =>
        {
            try
            {
                var channel = await EnsureChannelAsync(lifetime.ApplicationStopping);
                var body = JsonSerializer.SerializeToUtf8Bytes(message, JsonOptions);

                await channel.BasicPublishAsync(
                    exchange: options.Exchange,
                    routingKey: routingKey,
                    mandatory: false,
                    basicProperties: new BasicProperties
                    {
                        ContentType = "application/json",
                        DeliveryMode = DeliveryModes.Persistent
                    },
                    body: body,
                    cancellationToken: lifetime.ApplicationStopping);
            }
            catch (Exception exception)
            {
                await ResetAsync();
                logger.LogWarning("Klarte ikke å publisere {RoutingKey}: {Message}",
                    routingKey, exception.Message);
            }
        });
    }

    /// <summary>
    /// Kobler til ved behov og deklarerer exchangen hver gang en ny kanal lages. Da er
    /// exchangen garantert der før den første meldingen, også etter en reconnect.
    /// </summary>
    private async Task<IChannel> EnsureChannelAsync(CancellationToken cancellationToken)
    {
        if (_channel is { IsOpen: true })
        {
            return _channel;
        }

        await _gate.WaitAsync(cancellationToken);
        try
        {
            if (_channel is { IsOpen: true })
            {
                return _channel;
            }

            var factory = new ConnectionFactory
            {
                Uri = new Uri(options.Uri),
                AutomaticRecoveryEnabled = true
            };

            _connection = await factory.CreateConnectionAsync(cancellationToken);
            _channel = await _connection.CreateChannelAsync(cancellationToken: cancellationToken);
            await _channel.ExchangeDeclareAsync(options.Exchange, ExchangeType.Topic,
                durable: true, autoDelete: false, cancellationToken: cancellationToken);

            logger.LogInformation("Koblet til RabbitMQ. Exchange \"{Exchange}\" (topic, durable).",
                options.Exchange);

            return _channel;
        }
        finally
        {
            _gate.Release();
        }
    }

    private async Task ResetAsync()
    {
        try
        {
            if (_channel is not null)
            {
                await _channel.DisposeAsync();
            }

            if (_connection is not null)
            {
                await _connection.DisposeAsync();
            }
        }
        catch
        {
            // Broker kan allerede være borte. Det er greit.
        }
        finally
        {
            _channel = null;
            _connection = null;
        }
    }

    private static string Timestamp() =>
        DateTimeOffset.UtcNow.ToString("yyyy-MM-ddTHH:mm:ss.fffZ");

    public async ValueTask DisposeAsync()
    {
        await ResetAsync();
        _gate.Dispose();
    }
}
